File: android/src/main/java/expo/modules/arcgis/OfflineFunctions.kt
package expo.modules.arcgis

import com.arcgismaps.data.Geodatabase
import com.arcgismaps.mapping.PortalItem
import com.arcgismaps.portal.Portal
import com.arcgismaps.tasks.exportvectortiles.ExportVectorTilesTask
import com.arcgismaps.tasks.geodatabase.GeodatabaseSyncTask
import com.arcgismaps.tasks.geodatabase.SyncDirection
import com.arcgismaps.mapping.MobileMapPackage
import com.arcgismaps.tasks.offlinemaptask.OfflineMapSyncTask
import com.arcgismaps.tasks.offlinemaptask.OfflineMapTask
import com.arcgismaps.tasks.tilecache.ExportTileCacheTask
import com.arcgismaps.tasks.tilecache.EstimateTileCacheSizeJob
import expo.modules.kotlin.AppContext
import java.io.File
import kotlin.math.log2
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Free functions backing the JS `offline` namespace — take maps and data offline via the various
 * download/export tasks. Each is an async job that writes files under the app's storage.
 * Registered as `AsyncFunction`s in `ExpoArcgisGeometryModule` (the base dir is supplied there).
 */

/** Returns a fresh (emptied) download location under [baseDir]. Offline tasks need a non-existent dest. */
internal fun offlineDownloadDir(baseDir: File, name: String): File {
  val dir = File(baseDir, name)
  dir.deleteRecursively()
  return dir
}

/**
 * Returns a [JobRef] for an on-demand offline-map download (drive it via `result()` + `onProgress`).
 *
 * When [overrides] is non-null the function builds a `GenerateOfflineMapParameterOverrides` object
 * via `OfflineMapTask.createGenerateOfflineMapParameterOverrides`, trims the `levelIds` list on
 * every `ExportTileCacheParameters` entry to the subset that falls within [minScale, maxScale],
 * and passes the overrides object to `createGenerateOfflineMapJob`.
 *
 * **Vector-tile entries** (`ExportVectorTilesParameters`) expose only `maxLevel` (integer); the
 * SDK provides no minScale/maxScale setter on the override entry — deferred.
 */
internal suspend fun generateOfflineMap(
  appContext: AppContext,
  baseDir: File?,
  portalItemId: String,
  areaOfInterest: Map<String, Any?>,
  downloadName: String,
  overrides: Map<String, Any?>?,
): JobRef {
  val area = geometryFromDict(areaOfInterest) ?: throw IllegalArgumentException("Invalid area of interest")
  baseDir ?: throw IllegalStateException("No download directory available")
  val portalItem =
    PortalItem(Portal("https://www.arcgis.com", Portal.Connection.Anonymous), portalItemId)
  val task = OfflineMapTask(portalItem)
  val parameters = task.createDefaultGenerateOfflineMapParameters(area).getOrThrow()
  val dir = offlineDownloadDir(baseDir, downloadName)

  val paramOverrides = overrides?.let {
    val minScale = (it["minScale"] as? Number)?.toDouble() ?: 0.0
    val maxScale = (it["maxScale"] as? Number)?.toDouble() ?: 0.0
    val ovr = task.createGenerateOfflineMapParameterOverrides(parameters).getOrThrow()
    for ((_, tileCacheParams) in ovr.exportTileCacheParameters) {
      applyScaleToTileCacheParams(tileCacheParams.levelIds, minScale, maxScale)
    }
    // ExportVectorTilesParameters.maxLevel could narrow vector tiles but requires a level number,
    // not a scale denominator; left as-is (deferred).
    ovr
  }

  val job = task.createGenerateOfflineMapJob(parameters, dir.absolutePath, paramOverrides)
  return JobRef(appContext, job) {
    job.result().getOrThrow()
    mapOf("path" to dir.absolutePath)
  }
}

/**
 * Mutates [levelIds] in-place to keep only the levels within [minScale]..[maxScale].
 *
 * Level numbers are approximated from scale denominators using the standard Web Mercator formula:
 *   `level ≈ log2(591_657_550 / scale)` (valid for EPSG:3857 / WKID 102100 tile schemes).
 *
 * For services using a different LOD origin the cutoff will be approximate — see DEFER note.
 */
private fun applyScaleToTileCacheParams(
  levelIds: MutableList<Int>,
  minScale: Double,
  maxScale: Double,
) {
  if (levelIds.isEmpty() || (maxScale <= 0 && minScale <= 0)) return

  val level0Scale = 591_657_550.0

  fun scaleToLevel(scale: Double): Int {
    if (scale <= 0) return Int.MAX_VALUE
    return max(0, log2(level0Scale / scale).roundToInt())
  }

  val maxLevelForScale = if (maxScale > 0) scaleToLevel(maxScale) else Int.MAX_VALUE
  val minLevelForScale = if (minScale > 0) scaleToLevel(minScale) else 0

  levelIds.retainAll { id -> id in minLevelForScale..maxLevelForScale }
}

private fun offlineMapTask(portalItemId: String): OfflineMapTask =
  OfflineMapTask(PortalItem(Portal("https://www.arcgis.com", Portal.Connection.Anonymous), portalItemId))

internal suspend fun preplannedMapAreas(portalItemId: String): List<Map<String, Any?>> {
  val task = offlineMapTask(portalItemId)
  task.load().getOrThrow()
  return task.getPreplannedMapAreas().getOrThrow().mapIndexed { index, area ->
    area.load()
    mapOf("title" to area.portalItem.title, "index" to index)
  }
}

/**
 * Returns a [JobRef] that syncs a downloaded offline map (`.mmpk` at [mobileMapPackagePath]) with
 * its services — pushes local edits up and pulls server updates down.
 */
internal suspend fun syncOfflineMap(appContext: AppContext, mobileMapPackagePath: String): JobRef {
  val pkg = MobileMapPackage(mobileMapPackagePath)
  pkg.load().getOrThrow()
  val map = pkg.maps.firstOrNull() ?: throw IllegalArgumentException("No map in package")
  val task = OfflineMapSyncTask(map)
  val parameters = task.createDefaultOfflineMapSyncParameters().getOrThrow()
  val job = task.createOfflineMapSyncJob(parameters)
  return JobRef(appContext, job) {
    job.result().getOrThrow()
    mapOf("synced" to true)
  }
}

internal suspend fun downloadPreplannedOfflineMap(
  appContext: AppContext,
  baseDir: File?,
  portalItemId: String,
  areaIndex: Int,
  downloadName: String,
): JobRef {
  baseDir ?: throw IllegalStateException("No download directory available")
  val task = offlineMapTask(portalItemId)
  task.load().getOrThrow()
  val area = task.getPreplannedMapAreas().getOrThrow().getOrNull(areaIndex)
    ?: throw IllegalArgumentException("Invalid preplanned area index")
  val parameters = task.createDefaultDownloadPreplannedOfflineMapParameters(area).getOrThrow()
  val dir = offlineDownloadDir(baseDir, downloadName)
  val job = task.createDownloadPreplannedOfflineMapJob(parameters, dir.absolutePath)
  return JobRef(appContext, job) {
    job.result().getOrThrow()
    mapOf("path" to dir.absolutePath)
  }
}

internal suspend fun generateGeodatabase(
  appContext: AppContext,
  baseDir: File?,
  featureServiceUrl: String,
  extent: Map<String, Any?>,
  downloadName: String,
): JobRef {
  val area = geometryFromDict(extent) ?: throw IllegalArgumentException("Invalid extent")
  baseDir ?: throw IllegalStateException("No download directory available")
  val task = GeodatabaseSyncTask(featureServiceUrl)
  val parameters = task.createDefaultGenerateGeodatabaseParameters(area).getOrThrow()
  val file = offlineDownloadDir(baseDir, "$downloadName.geodatabase")
  val job = task.createGenerateGeodatabaseJob(parameters, file.absolutePath)
  return JobRef(appContext, job) {
    val geodatabase = job.result().getOrThrow()
    mapOf("path" to geodatabase.path, "tableCount" to geodatabase.featureTables.size)
  }
}

internal suspend fun syncGeodatabase(
  appContext: AppContext,
  geodatabasePath: String,
  featureServiceUrl: String,
): JobRef {
  val geodatabase = Geodatabase(geodatabasePath)
  geodatabase.load().getOrThrow()
  val task = GeodatabaseSyncTask(featureServiceUrl)
  val job = task.createSyncGeodatabaseJob(SyncDirection.Bidirectional, true, geodatabase)
  return JobRef(appContext, job) {
    job.result().getOrThrow()
    mapOf("synced" to true)
  }
}

internal suspend fun exportTileCache(
  appContext: AppContext,
  baseDir: File?,
  tileServiceUrl: String,
  areaOfInterest: Map<String, Any?>,
  downloadName: String,
): JobRef {
  val area = geometryFromDict(areaOfInterest) ?: throw IllegalArgumentException("Invalid area of interest")
  baseDir ?: throw IllegalStateException("No download directory available")
  val task = ExportTileCacheTask(tileServiceUrl)
  val parameters = task.createDefaultExportTileCacheParameters(area, 0.0, 0.0).getOrThrow()
  val file = offlineDownloadDir(baseDir, "$downloadName.tpkx")
  val job = task.createExportTileCacheJob(parameters, file.absolutePath)
  return JobRef(appContext, job) {
    job.result().getOrThrow()
    mapOf("path" to file.absolutePath)
  }
}

/**
 * Estimates the on-disk file size and tile count for an [exportTileCache] download WITHOUT
 * downloading. Passes [minScale]/[maxScale] (0.0 = no limit) to
 * [ExportTileCacheTask.createDefaultExportTileCacheParameters] to narrow the estimation to the
 * same scale window that [exportTileCache] would use. Runs the [EstimateTileCacheSizeJob] to
 * completion and returns `{fileSizeBytes, tileCount}`.
 */
internal suspend fun estimateTileCacheSize(
  tileServiceUrl: String,
  areaOfInterest: Map<String, Any?>,
  minScale: Double?,
  maxScale: Double?,
): Map<String, Any?> {
  val area = geometryFromDict(areaOfInterest) ?: throw IllegalArgumentException("Invalid area of interest")
  val task = ExportTileCacheTask(tileServiceUrl)
  val parameters = task.createDefaultExportTileCacheParameters(
    area,
    minScale ?: 0.0,
    maxScale ?: 0.0,
  ).getOrThrow()
  val job = task.createEstimateTileCacheSizeJob(parameters)
  job.start()
  val estimate = job.result().getOrThrow()
  return mapOf("fileSizeBytes" to estimate.fileSize.toDouble(), "tileCount" to estimate.tileCount.toDouble())
}

internal suspend fun exportVectorTiles(
  appContext: AppContext,
  baseDir: File?,
  vectorTileServiceUrl: String,
  areaOfInterest: Map<String, Any?>,
  downloadName: String,
): JobRef {
  val area = geometryFromDict(areaOfInterest) ?: throw IllegalArgumentException("Invalid area of interest")
  baseDir ?: throw IllegalStateException("No download directory available")
  val task = ExportVectorTilesTask(vectorTileServiceUrl)
  val parameters = task.createDefaultExportVectorTilesParameters(area, 0.0).getOrThrow()
  val file = offlineDownloadDir(baseDir, "$downloadName.vtpk")
  val job = task.createExportVectorTilesJob(parameters, file.absolutePath)
  return JobRef(appContext, job) {
    job.result().getOrThrow()
    mapOf("path" to file.absolutePath)
  }
}

File: ios/OfflineFunctions.swift
import ArcGIS
import Foundation

/// Free functions backing the JS `offline` namespace — take maps and data offline via the various
/// download/export tasks. Each is an async job that writes files into the app's documents dir.
/// Registered as `AsyncFunction`s in `ExpoArcgisGeometryModule`.

/// Returns a fresh (emptied) download location under the app's documents directory. The offline
/// tasks require the destination not to already exist.
func offlineDownloadURL(_ name: String) -> URL {
  let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
  let url = documents.appendingPathComponent(name)
  try? FileManager.default.removeItem(at: url)
  return url
}

enum OfflineError: Error { case invalidParameters }

/// Returns a `JobRef` for an on-demand offline-map download (drive it via `result()` + `onProgress`).
/// When `overrides` contains `minScale`/`maxScale`, the function builds a
/// `GenerateOfflineMapParameterOverrides` object, trims the `levelIDs` on every
/// `ExportTileCacheParameters` entry to the scale window, then passes the overrides object to the
/// job constructor.
///
/// **Vector-tile entries** (`ExportVectorTilesParameters`) are not filtered — `SDK 300.0` exposes
/// no per-entry scale setter on `ExportVectorTilesParameters` via the override map (deferred).
func generateOfflineMap(
  _ portalItemId: String,
  _ areaOfInterest: [String: Any],
  _ downloadName: String,
  _ overridesDict: [String: Any]?
) async throws -> JobRef {
  guard let area = geometryFromDict(areaOfInterest), let itemID = PortalItem.ID(portalItemId) else {
    throw OfflineError.invalidParameters
  }
  let portalItem = PortalItem(portal: .arcGISOnline(connection: .anonymous), id: itemID)
  let task = OfflineMapTask(portalItem: portalItem)
  let parameters = try await task.makeDefaultGenerateOfflineMapParameters(areaOfInterest: area)
  let directory = offlineDownloadURL(downloadName)

  // Build parameter overrides when the caller supplied an overrides dictionary.
  var paramOverrides: GenerateOfflineMapParameterOverrides?
  if let dict = overridesDict {
    let ovr = try await task.makeGenerateOfflineMapParameterOverrides(parameters: parameters)
    let minScale = dict["minScale"] as? Double ?? 0
    let maxScale = dict["maxScale"] as? Double ?? 0
    // Apply scale window to every ExportTileCacheParameters entry.
    for (key, tileCacheParams) in ovr.exportTileCacheParameters {
      applyScaleToTileCacheParams(tileCacheParams, minScale: minScale, maxScale: maxScale)
      ovr.setExportTileCacheParameterValue(tileCacheParams, forKey: key)
    }
    // Note: ExportVectorTilesParameters has no per-entry scale filter in SDK 300.0 — deferred.
    paramOverrides = ovr
  }

  let job = task.makeGenerateOfflineMapJob(
    parameters: parameters,
    downloadDirectory: directory,
    overrides: paramOverrides
  )
  return JobRef(job: job) {
    _ = try await job.result.get()
    return ["path": directory.path]
  }
}

/// Trims `ExportTileCacheParameters.levelIDs` to the subset within [minScale, maxScale].
///
/// Level IDs in ArcGIS tile caches are WMS/WMTS LOD indices (0 = world, N = street-level).
/// We approximate the level number for a given scale using the standard Web Mercator formula:
///   `level ≈ log2(591_657_550 / scale)` (valid for WKID 3857 / 102100 tile schemes).
///
/// For services that use a different LOD scheme (e.g. 1:1 000 000 at level 0) the cutoff
/// will be approximate. This is an accepted approximation — see DEFER note in the caller.
///
/// - `maxScale > 0`: drop level IDs strictly finer than `maxScaleLevel` (i.e. id > maxScaleLevel).
/// - `minScale > 0`: drop level IDs strictly coarser than `minScaleLevel` (i.e. id < minScaleLevel).
private func applyScaleToTileCacheParams(
  _ params: ExportTileCacheParameters,
  minScale: Double,
  maxScale: Double
) {
  let levels = params.levelIDs  // read-only; sorted ascending
  guard !levels.isEmpty, maxScale > 0 || minScale > 0 else { return }

  // Reference scale denominator for level 0 in the standard Web Mercator tile scheme.
  let level0Scale: Double = 591_657_550

  func scaleToLevel(_ scale: Double) -> Int {
    guard scale > 0 else { return Int.max }
    let lvl = log2(level0Scale / scale)
    return max(0, Int(lvl.rounded()))
  }

  let maxLevelForScale = maxScale > 0 ? scaleToLevel(maxScale) : Int.max
  let minLevelForScale = minScale > 0 ? scaleToLevel(minScale) : 0

  let filtered = levels.filter { id in id >= minLevelForScale && id <= maxLevelForScale }
  guard filtered != levels else { return }
  params.removeAllLevelIDs()
  params.addLevelIDs(filtered)
}

private func offlineMapTask(_ portalItemId: String) -> OfflineMapTask? {
  guard let itemID = PortalItem.ID(portalItemId) else { return nil }
  return OfflineMapTask(portalItem: PortalItem(portal: .arcGISOnline(connection: .anonymous), id: itemID))
}

func preplannedMapAreas(_ portalItemId: String) async throws -> [[String: Any]] {
  guard let task = offlineMapTask(portalItemId) else { return [] }
  try await task.load()
  var result: [[String: Any]] = []
  for (index, area) in try await task.preplannedMapAreas.enumerated() {
    try? await area.load()
    result.append(["title": area.portalItem.title, "index": index])
  }
  return result
}

func downloadPreplannedOfflineMap(_ portalItemId: String, _ areaIndex: Int, _ downloadName: String) async throws -> JobRef {
  guard let task = offlineMapTask(portalItemId) else { throw OfflineError.invalidParameters }
  try await task.load()
  let areas = try await task.preplannedMapAreas
  guard areaIndex >= 0, areaIndex < areas.count else { throw OfflineError.invalidParameters }
  let parameters = try await task.makeDefaultDownloadPreplannedOfflineMapParameters(preplannedMapArea: areas[areaIndex])
  let directory = offlineDownloadURL(downloadName)
  let job = task.makeDownloadPreplannedOfflineMapJob(parameters: parameters, downloadDirectory: directory)
  return JobRef(job: job) {
    _ = try await job.result.get()
    return ["path": directory.path]
  }
}

/// Returns a `JobRef` that syncs a downloaded offline map (`.mmpk` at `mobileMapPackagePath`) with
/// its services — pushes local edits up and pulls server updates down.
func syncOfflineMap(_ mobileMapPackagePath: String) async throws -> JobRef {
  let package = MobileMapPackage(fileURL: URL(fileURLWithPath: mobileMapPackagePath))
  try await package.load()
  guard let map = package.maps.first else { throw OfflineError.invalidParameters }
  let task = OfflineMapSyncTask(map: map)
  let parameters = try await task.makeDefaultOfflineMapSyncParameters()
  let job = task.makeSyncOfflineMapJob(parameters: parameters)
  return JobRef(job: job) {
    _ = try await job.result.get()
    return ["synced": true]
  }
}

func generateGeodatabase(_ featureServiceUrl: String, _ extent: [String: Any], _ downloadName: String) async throws -> JobRef {
  guard let url = URL(string: featureServiceUrl), let area = geometryFromDict(extent) else {
    throw OfflineError.invalidParameters
  }
  let task = GeodatabaseSyncTask(url: url)
  let parameters = try await task.makeDefaultGenerateGeodatabaseParameters(extent: area)
  let fileURL = offlineDownloadURL(downloadName + ".geodatabase")
  let job = task.makeGenerateGeodatabaseJob(parameters: parameters, downloadFileURL: fileURL)
  return JobRef(job: job) {
    let geodatabase = try await job.result.get()
    return ["path": geodatabase.fileURL.path, "tableCount": geodatabase.featureTables.count]
  }
}

func syncGeodatabase(_ geodatabasePath: String, _ featureServiceUrl: String) async throws -> JobRef {
  guard let url = URL(string: featureServiceUrl) else { throw OfflineError.invalidParameters }
  let geodatabase = Geodatabase(fileURL: URL(fileURLWithPath: geodatabasePath))
  try await geodatabase.load()
  let task = GeodatabaseSyncTask(url: url)
  let job = task.makeSyncGeodatabaseJob(syncDirection: .bidirectional, rollbackOnFailure: true, geodatabase: geodatabase)
  return JobRef(job: job) {
    _ = try await job.result.get()
    return ["synced": true]
  }
}

func exportTileCache(_ tileServiceUrl: String, _ areaOfInterest: [String: Any], _ downloadName: String) async throws -> JobRef {
  guard let url = URL(string: tileServiceUrl), let area = geometryFromDict(areaOfInterest) else {
    throw OfflineError.invalidParameters
  }
  let task = ExportTileCacheTask(url: url)
  let parameters = try await task.makeDefaultExportTileCacheParameters(areaOfInterest: area)
  let fileURL = offlineDownloadURL(downloadName + ".tpkx")
  let job = task.makeExportTileCacheJob(parameters: parameters, downloadFileURL: fileURL)
  return JobRef(job: job) {
    _ = try await job.result.get()
    return ["path": fileURL.path]
  }
}

/// Estimates the on-disk file size and tile count for an `exportTileCache` download WITHOUT
/// downloading. Runs the `EstimateTileCacheSizeJob` to completion and returns `{fileSize, tileCount}`.
func estimateTileCacheSize(_ tileServiceUrl: String, _ areaOfInterest: [String: Any], _ options: [String: Any]?) async throws -> [String: Any] {
  guard let url = URL(string: tileServiceUrl), let area = geometryFromDict(areaOfInterest) else {
    throw OfflineError.invalidParameters
  }
  let task = ExportTileCacheTask(url: url)
  let parameters = try await task.makeDefaultExportTileCacheParameters(areaOfInterest: area)
  let job = task.makeEstimateTileCacheSizeJob(parameters: parameters)
  job.start()
  let estimate = try await job.result.get()
  return ["fileSize": estimate.fileSize, "tileCount": estimate.tileCount]
}

func exportVectorTiles(_ vectorTileServiceUrl: String, _ areaOfInterest: [String: Any], _ downloadName: String) async throws -> JobRef {
  guard let url = URL(string: vectorTileServiceUrl), let area = geometryFromDict(areaOfInterest) else {
    throw OfflineError.invalidParameters
  }
  let task = ExportVectorTilesTask(url: url)
  let parameters = try await task.makeDefaultExportVectorTilesParameters(areaOfInterest: area)
  let fileURL = offlineDownloadURL(downloadName + ".vtpk")
  let job = task.makeExportVectorTilesJob(parameters: parameters, downloadFileURL: fileURL)
  return JobRef(job: job) {
    _ = try await job.result.get()
    return ["path": fileURL.path]
  }
}
